pkg/core/sync.go

```diff
@@ -0,0 +1,258 @@
+	return res, diags
+}
+
+func Drop(ctx context.Context, storage database.Storage, pm *plugin.Manager, provider registry.Provider) diag.Diagnostics {
+	log.Warn().Stringer("provider", provider).Msg("dropping provider schema")
+	s, diags := GetProviderSchema(ctx, pm, &GetProviderSchemaOptions{Provider: provider})
+	if diags.HasDiags() {
+		return diags
+	}
+
+	sta, err := state.NewClient(ctx, storage.DSN())
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer sta.Close()
+
+	tx, err := sta.ProviderSync(ctx)
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer tx.Rollback(ctx) // nolint:errcheck
+
+	log.Info().Str("provider", provider.Name).Str("version", provider.Version).Msg("dropping provider tables")
+	if diags := dropProvider(ctx, tx, provider, resourceTableNames(s.ResourceTables), nil, nil); diags.HasErrors() {
+		return diags
+	}
+
+	if err := tx.UninstallProvider(ctx, provider); err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return diag.FromError(err, diag.DATABASE)
+	}
+
+	return nil
+}
+
+func syncTables(ctx context.Context, sta *state.Client, cur, want *state.Provider, resourceTables map[string]*schema.Table) diag.Diagnostics {
+	want.Tables = resourceTableNames(resourceTables)
+	want.Signatures = resourceSignatures(resourceTables)
+
+	var (
+		curTables     map[string][]string
+		curSignatures map[string]string
+	)
+
+	if cur != nil && len(cur.Tables) > 0 { // Old provider with known, valid data
+		curTables = cur.Tables
+		curSignatures = cur.Signatures
+	} else {
+		curTables = want.Tables // Fallback to installed provider tables
+	}
+
+	tx, err := sta.ProviderSync(ctx)
+	if err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+	defer tx.Rollback(ctx) // nolint:errcheck
+
+	// If a single SQL fails, all following commands also fail with "current transaction is aborted"
+	if diags := dropProvider(ctx, tx, want.Registry(), curTables, curSignatures, want.Signatures); diags.HasErrors() {
+		return diags
+	}
+	if diags := installProvider(ctx, tx, resourceTables); diags.HasErrors() {
+		return diags
+	}
+
+	if err := tx.UninstallProvider(ctx, want.Registry()); err != nil {
+		return diag.FromError(fmt.Errorf("uninstall provider failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.InstallProvider(ctx, want); err != nil {
+		return diag.FromError(fmt.Errorf("state failed: %w", err), diag.INTERNAL)
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return diag.FromError(err, diag.DATABASE)
+	}
+
+	return nil
+}
+
+func dropProvider(ctx context.Context, db execution.QueryExecer, provider registry.Provider, tableNames map[string][]string, curSignatures, wantSignatures map[string]string) diag.Diagnostics {
+	{
```

Comment:
nit: why the brackets?

Reply:
scoping. it's a block, migTable and q isn't used anywhere else, which can be complicated, so it eases reading.